blocks/search.py

```diff
@@ -0,0 +1,269 @@
+                       for state in next_values]
+        return self.generate_names + ['probs'], next_values
+
+    @classmethod
+    def _top_probs(cls, probs, beam_size, unique=False):
+        """Returns indexes of elements with highest probabilities.
+
+        Parameters
+        ----------
+        probs : numpy array
+            A 3d array of probabilities (length of sequence, batch,
+            readout_dim)
+        beam_size : int
+            Beam size, number of top probs to return
+
+        Returns
+        -------
+        Tuple of (indexes, top probabilities)
+
+        """
+        flatten = probs.flatten()
+        if unique:
+            args = numpy.unique(
+                numpy.argpartition(-flatten, beam_size))[:beam_size]
+        else:
+            args = numpy.argpartition(-flatten, beam_size)[:beam_size]
+        args = args[numpy.argsort(-flatten[args])]
+        if unique:
+            # append best if needed
+            if args.shape[0] < beam_size:
+                args = numpy.append(
+                    args,
+                    numpy.tile(args[0], beam_size - args.shape[0]))
+        # convert args back
+        indexes = numpy.unravel_index(args, probs.shape[1:])
+        return indexes, probs[0][indexes]
+
+    @staticmethod
+    def _rearrange(outputs, indexes):
+        new_outputs = outputs[:, indexes.T.flatten()]
+        return new_outputs.copy()
+
+    def search(self, inputs_val_dict, eol_symbol=-1, max_length=512):
+        """Performs beam search.
+
+        Parameters
+        ----------
+        inputs_val_dict : dict
+            Dictionary of input values {name: value}. Input values may
+            be in a batch of size greater than 1, in this case, several
+            beam searches will be performed in parallel.
+        eol_symbol : int
+            End of sequence symbol, the search stops when the
+            symbol is generated.
+        max_length : int
+            Maximum sequence length, the search stops when it
+            is reached.
+
+        Returns
+        -------
+        Most probable sequences, corresponding probabilities and costs.
+
+        """
+        if not self.compiled:
+            self.compile()
+        # Inputs repeated beam_size times
+        aux_inputs = OrderedDict(
+            [(name, numpy.tile(val, (1, self.beam_size)))
+             for name, val in inputs_val_dict.iteritems()])
+
+        # Precompute contexts
+        contexts = self.compute_contexts(aux_inputs)
+
+        cur_states = self.compute_initial_states(contexts)
+
+        cur_states['cur_outputs'] = cur_states['outputs']
+        cur_states['cur_outputs_mask'] = numpy.ones_like(
+            cur_states['cur_outputs'])
+        cur_states['cur_probs'] = numpy.ones_like(
+            cur_states['cur_outputs'])
+
+        for i in range(max_length):
+            cur_states.update(self.compute_next(contexts, cur_states))
+            next_probs = (cur_states['cur_probs'][:, :, None] *
```

Comment:
Please also check it the `top_probs` documentation is up-to-date. I am not sure about that.